torch/_prims/__init__.py

```diff
@@ -0,0 +1,992 @@
+#
+
+# Describes the return type of the primitive:
+#
+#   - NEW, a new tensor is created
+#   - VIEW, a view of an input tensor is returned
+#   - INPLACE, one or more input tensors is modified
+#
+# these descriptors are mututally exclusive and exhaustive.
+class RETURN_TYPE(Enum):
+    NEW = (0,)
+    VIEW = (1,)
+    INPLACE = (2,)
+
+
+def _make_prim(
+    *, meta: Callable, impl_aten: Callable, return_type: RETURN_TYPE, doc: str
+):
+    """
+    Creates a primitive operation.
+
+    """
+
+    def _prim(*args, **kwargs):
+        # TODO: refactor this to not incur such a latency hit
+        return_meta = False
+        for arg in args:
+            if isinstance(arg, TensorMeta):
+                return_meta = True
+        for v in kwargs.values():
+            if isinstance(v, TensorMeta):
+                return_meta = True
+        meta_result = meta(*args, **kwargs)
+
+        if return_meta:
+            return meta_result
+
+        return impl_aten(*args, **kwargs)
+
+    _prim.__doc__ = doc
+    _prim.meta = meta
+    _prim.return_type = return_type
+
+    return _prim
+
+
+def _elementwise_meta(*args):
+    '''
+    Meta function for elementwise operations.
+
+    Stride logic is currently incorrect.
+    '''
+
+    assert len(args) > 0
+
+    utils.check_same_device(*args, allow_scalars=True)
+    utils.check_same_shape(*args)
+    utils.check_same_dtype(*args)
+
+    strides = None
+    tensor = None
+    number = None
+    for arg in args:
+        if isinstance(arg, TensorLike):
+            if strides is None:
+                strides = arg.stride()
+
+            if tensor is None:
+                tensor = arg
+
+            if arg.stride() != strides:
+                return TensorMeta(
+                    arg, strides=utils.make_contiguous_strides_for(arg.shape)
+                )
+        elif isinstance(arg, Number):
+            if number is None:
+                number = arg
+
+    # TODO: fix strides
+    if tensor is not None:
+        if 0 in tensor.stride() and tensor.numel() > 0:
+            return TensorMeta(tensor, strides=utils.make_contiguous_strides_for(tensor.shape))
+        else:
+            return TensorMeta(tensor)
+
+    return TensorMeta(number)
+
+
+def _make_elementwise_unary_prim(impl_aten: Callable, doc: str):
+    """
+    Creates an elementwise unary prim.
+    """
+
+    return _make_prim(
+        meta=_elementwise_meta,
+        impl_aten=impl_aten,
+        return_type=RETURN_TYPE.NEW,
+        doc=doc,
+    )
+
+
+def _make_elementwise_binary_prim(impl_aten: Callable, doc: str):
+    """
+    Creates an elementwise binary prim.
+    """
+
+    return _make_prim(
+        meta=_elementwise_meta,
+        impl_aten=impl_aten,
+        return_type=RETURN_TYPE.NEW,
+        doc=doc,
+    )
+
+
+def _not_impl(*args, **kwargs):
+    raise NotImplementedError
+
+
+#
+# Elementwise unary operations
+#
+
+abs = _make_elementwise_unary_prim(
+    impl_aten=torch.abs,
+    doc="",
+)
```

Comment:
So... do you really want this to follow aten semantics for complex numbers?

Reply:
@anjali411 is considering this

Reply:
What is the concern regarding ATen complex semantics?

Reply:
What is the concern regarding ATen complex semantics?